packages/web/src/components/createCell.tsx

```diff
@@ -1,16 +1,48 @@
+import { useMemo } from 'react'
+
 import type { DocumentNode } from 'graphql'
 import type { A } from 'ts-toolbelt'
 
 import { useQuery } from './GraphQLHooksProvider'
 
+enum Status {
+  Error = 'ERROR',
+  Loading = 'LOADING',
+  Success = 'SUCCESS',
+  Updating = 'UPDATING',
+  Empty = 'EMPTY',
+  NullDataError = 'NULL_DATA_ERROR',
+}
+
 interface QueryProps {
   query: DocumentNode
-  children: (result: QueryOperationResult) => React.ReactElement
+  children: (
+    result: QueryOperationResult & { status: Status }
+  ) => React.ReactElement
 }
 
 const Query = ({ children, query, ...rest }: QueryProps) => {
   const result = useQuery(query, rest)
-  return result ? children(result) : null
+
+  const status = useMemo(() => {
+    if (result.error) {
+      return Status.Error
+    } else if (result.data) {
+      if (result.loading) {
+        return Status.Updating
+      } else if (isEmpty(result.data)) {
+        return Status.Empty
+      } else {
+        return Status.Success
+      }
+    } else if (result.loading) {
+      return Status.Loading
+    } else {
+      return Status.NullDataError
+    }
+  }, [result])
```

Comment:
The logic doesn't feel any better yet. Still trying to figure out if I need to use a combination of `useReducer` and `useEffect` or something like that.

Reply:
I don't think `useMemo` is making this code any faster or easier to reason about, it's probably making it slower, because useMemo is doing more work than just computing the "status:"

https://kentcdodds.com/blog/usememo-and-usecallback/

Reply:
And, I don't know if abstracting the "response status" to determine what to render is helpful, since lower down we need additional things to consider if we should render that status, and now you need to keep two pieces of code and logic in your brain to determine if the conditions for rendering things are met.

My suggestion would be to return early if the correct conditions are met.

```
const { error } = response

if (error) {
    if (!Errror) {
       console.warn("You have not exported "Error" on your Cell, therefore this error is unhandled.")
       throw error
    }
    return <Error error={error} />
}
```